More consistency between UniformFanInShape and UniformFanOutShape

laszlovandenhoek opened the pull request:
`UniformFanInShape` has `inSeq`, whereas `UniformFanOutShape` has `outArray`.
There's probably no good reason for this discrepancy, so correct it.

laszlovandenhoek committed:
add+use UniformFanOutShape#outSeq for consistency

UniformFanInShape has inSeq, whereas UniformFanOutShape has outArray.
There's probably no good reason for this discrepancy.

ktoso commented:
Hmm, I think I'd rather want to stick to the array version, should have less overhead than the Seq.
If not sure we could benchmark for it - selecting outputs in large shapes.
Otherwise good idea to make them consistent, yeah.

johanandren commented on akka-stream/src/main/boilerplate/akka/stream/FanOutShape.scala.template:
```diff
@@ -59,8 +59,10 @@ class UniformFanOutShape[-I, +O](n: Int, _init: FanOutShape.Init[I @uncheckedVar
-  def out(n: Int): Outlet[O @uncheckedVariance] = outArray(n)
+  val outSeq: immutable.IndexedSeq[Outlet[O @uncheckedVariance]] = Vector.tabulate(n)(i => newOutlet[O](s"out$i"))
+  @deprecated("outArray is deprecated; use outSeq instead", "2.5.3")
+  val outArray: Array[Outlet[O @uncheckedVariance]] = outSeq.toArray
```
We should rather have `outSeq` method (or field) casting the array than having a `Vector`, replacing the data structure will have performance implications .

johanandren commented:
Fixes #23322

johanandren commented:
OK TO TEST

laszlovandenhoek commented:
@ktoso @johanandren since we're exposing the object, shouldn't we be preferring an immutable `Seq` over a mutable `Array ?

ktoso commented:
The exposing may have been accidental... The way people are supposed to use these is to call ` def out(n: Int): Outlet[O @uncheckedVariance] = outArray(n)`

laszlovandenhoek committed:
2.5.3 is gold, deprecation would start in 2.5.4

johanandren commented:
`Vector` has much worse performance and allocation overhead. (See Li's benchmarks for example: http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html)

That it is publicly mutable is likely a mistake, but that it is an array is an implementation detail that may be important.

laszlovandenhoek commented:
@ktoso I still have a use case for accessing the underlying collection; I want to be able to call `last` like this (last line):

```
  private def fetchUntilLast(implicit system: ActorSystem, materializer: Materializer) = GraphDSL.create() { implicit builder =>
    import GraphDSL.Implicits._

    val join = builder.add(Merge[Uri](2))
    val split = builder.add(Broadcast[HttpResponse](2))

    join ~> fetchUri ~> split
    join <~ nextPage <~ split

    FlowShape(join.inSeq.last, split.outArray.last)
  }
```

The way to use this is to plug in an initial `Uri` at the first join, fetch it, spit out subsequent `HttpResponse`s while they come out and in parallel extract the HATEOAS `next` header to feed the next page of the paginated result back into the Stream.

I'm still not entirely happy with this and perhaps there already exists a better way to do this, but I couldn't find it.

laszlovandenhoek commented:
@johanandren I took this code from `UniformFanInShape`, and `Vector` seems to be working fine there, so I can't think of a way how it being an `Array` here could be important (other than from a performance perspective). The only two places that use the `outArray` field directly do so to access its `length`.

johanandren commented:
No it isn't the only place, `akka.stream.UniformFanOutShape#out` does a lookup every time. As per the benchmarks I linked to, Vector lookup in a small sized collection is an order of magnitude slower, for every lookup, everywhere `out(n)` is used in user code. Or causes an extra copy of the collection for every use of the shape with the current change.

ktoso commented:
>     FlowShape(join.inSeq.last, split.outArray.last)

I would not count this as reason enough to maintain what actually is a leaked implementation detail.
The same is doable with `out(n - 1)` indexing.

So to address these I would suggest moving to arrays everywhere, perhaps deprecating the array field so people do not touch them directly and suggesting there to use the out method.

ktoso reviewed (changes requested):
>     FlowShape(join.inSeq.last, split.outArray.last)

I would not count this as reason enough to maintain what actually is a leaked implementation detail.
The same is doable with `out(n - 1)` indexing.

So to address these I would suggest moving to arrays everywhere, perhaps deprecating the array field so people do not touch them directly and suggesting there to use the out method.

laszlovandenhoek commented:
Alright guys, thanks for the feedback, very educational :) Back to the drawing board it is! Will get back with something.

ktoso commented:
Great, thanks in advance!

johanandren commented:
I'd think a use case for a collection out outs is where you want to iterate over the outputs. @ktoso you don't think that is common/useful?

laszlovandenhoek commented:
There is at least one place where `UniformFanInShape.inSeq` is used, namely in `class ZipWithN` (Graph.scala:964 to be precise). Moreover, it is explicitly assumed to be an `immutable.Seq` there (Graph.scala:975). Hmm.

Maybe it would be OK after all to take the overhead of creating an immutable Seq, for the sake of a more uniform API? After all, `UniformFanXShape` ultimately extends from `Shape`, which (quoting from its ScalaDoc) "_describes the inlets and outlets of a *Graph*_", and "_a Graph is a freely reusable blueprint_", so you really shouldn't need to instantiate it more than once anyway.

johanandren commented:
Not really, everywhere it is used it is created with that set of inputs, through the `apply` method. So everywhere someone has created a custom fan in `GraphStage` they will instantiate a new shape instance.

A few realisations:
 * the base class actually always keeps a vector in `_outlets`, so we are already maintaining a copy, and that seems like a waste. In the `Array` case that could be explained by being an optimisation, but the reason to keep an identical `Vector` I cannot see.
 * We actually even have samples of how to extend `FanInShape` this means that if we add a method, a user could have already created a method with that name with a different signature - I wonder if this does not stop use from doing anything here except possibly deprecating.

laszlovandenhoek commented:
@johanandren I considered your remarks w.r.t. `_outlets` and `_inlets` already being added by `Fan...Shape`, and only then did it hit me what distinguishes `UniformFan...Shape` from its parent `Fan...Shape`, namely: all inputs are of the same type.

It would be easiest if we could just use `inlets` and `outlets` from `Shape`; unfortunately, these are typed as `immutable.Seq[Inlet[_]]` and `immutable.Seq[Outlet[_]]`, respectively. I can't specialize these as `immutable.Seq[Inlet[I]]` and `immutable.Seq[Outlet[O]]` in `UniformFan...Shape`, because they're declared `final` in `Fan...Shape`, because they're implemented there.

I feel it would be best to rename the existing `Fan...Shape` types to `AmorphousFan...Shape` or `PluriformFan...Shape` or NonUniformFan...Shape` or something, pulling up the shared details to a new `AbstractFan...Shape`, leaving `UniformFan...Shape` free to specialize the `inputs` and `outputs` with the correct type, instead of `_`.

If we don't change the name of `FanInShape` this wouldn't need to be a breaking change. What do you think?

laszlovandenhoek commented:
@johanandren I guess binary compatibility would be broken if `UniformFan...Shape` no longer extended `Fan...Shape`, so we can't introduce a new, `private[stream]` common parent that encapsulates the `_inlets` and `_outlets` to prevent extenders from accidentally duplicating those fields, which I think is the whole point of having `def inlets` and `def outlets` be `final`. That might be something for 2.6 though, I guess.

For now, let me stick to specializing `UniformFan...Shape.[in|out]lets` and deprecating the extra fields.

laszlovandenhoek committed:
specialize in/outlets where possible

laszlovandenhoek commented:
Alright, done for now. `FanInShape1N` remains to be done. I'll wait for approval on the work so far first, though.

johanandren commented:
Let's see what MiMa says, if you don't want to wait on our CI server you can also run it locally using `akka-stream/mimaReportBinaryIssues`

laszlovandenhoek commented:
Alright, what seems to fail is a boatload of instances of "Cannot pull port (MergePreferred.preferred(_nnn_)) twice", all for `FanInShape`-related stuff (no problems with `FanOutShape`). This is somewhat puzzling but I guess I'll just have to run the tests locally and see what goes wrong.

No unsurmountable binary issues reported, though :)

ktoso reviewed (commented):
Needs more polish. Originally I was thinking to remain with the Array internally and go to Seq only when needed - on the deprecated methods. Have you attempted it that way? We would not want people to touch that array, sure, but for implementing the `in(i: Int)` an array is the best thing so I'd perhaps rather do that than start with Vector and then toArray

laszlovandenhoek committed:
review comments, binary compatibility

laszlovandenhoek committed:
optimize imports

laszlovandenhoek commented on akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template:
```diff
@@ -53,15 +53,25 @@ object UniformFanInShape {
+
+  @deprecated("use inlets or in(n) instead", "2.5.4")
+  val inSeq: immutable.Seq[Inlet[T @uncheckedVariance]] = inlets
+  def in(n: Int): Inlet[T @uncheckedVariance] = inlets(n)
```
Nope, binary compatibility check says no :(

laszlovandenhoek commented:
@ktoso thanks for the review, I made the suggested changes (modulo binary compatibility constraints).

I understand that for lookups, as in `in(n)`, an `Array` would be most efficient. But `FanInShape` already has `_inlets` internally, so keeping them in a separate `Array` in `UniformFanInShape` would be a duplication, as @johanandren also remarked. So I set out to eliminate the extra collection within the `UniformFanXShape` classes. Unfortunately, this resulted in those failing unit tests, and I still haven't figured out why. I'll make a (less ambitious) attempt with an `Array` in `UniformFanInShape`.

laszlovandenhoek committed:
use Array copy internally

laszlovandenhoek committed:
give FanInShape1N the deprecation treatment

also, s/T\d/I\d/g

laszlovandenhoek commented:
Following @ktoso 's suggestion to use an `Array` internally fixes the tests. Go figure...

The end result is that the (non-deprecated) public API is more uniform now, but the implementations of `UniformFanInShape` and `UniformFanOutShape` have actually diverged. I hope this is OK.

laszlovandenhoek committed:
delete ignored file

ktoso commented:
> Following @ktoso 's suggestion to use an Array internally fixes the tests. Go figure...

😉 

> The end result is that the (non-deprecated) public API is more uniform now, but the implementations of UniformFanInShape and UniformFanOutShape have actually diverged. I hope this is OK.

Yes, I think that's fine. Internally we often have to take a hit and do things not nice, in order to give users a nice API.

ktoso commented on akka-stream/src/main/boilerplate/akka/stream/FanOutShape.scala.template:
```diff
@@ -53,14 +53,20 @@ object UniformFanOutShape {
+  final override def outlets: immutable.Seq[Outlet[O @uncheckedVariance]] = super.outlets.asInstanceOf[immutable.Seq[Outlet[O]]]
+
+  @deprecated("use `outlets` or `out(id)` instead", "2.5.4")
+  lazy val outArray: Array[Outlet[O @uncheckedVariance]] = outlets.toArray
```
Since this one, unlike the Seq may have been used by Java people, we should annotate it as Java's Deprecated as well. See here: https://github.com/akka/akka/pull/21913/files#diff-ac833d777f0270e843a0ed2af8aae769R23 This is to make sure it also shows up as deprecated for Java users without any scala plugins etc

ktoso reviewed (commented):
I think this is almost there, another pair of eyes on it is needed though so we don't accidentally break things here. Thanks for your continued effort!

laszlovandenhoek commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -215,7 +215,7 @@ final class MergePreferred[T](val secondaryPorts: Int, val eagerComplete: Boolea
     override def preStart(): Unit = {
       tryPull(preferred)
-      shape.inSeq.foreach(tryPull)
+      shape.inArray.foreach(tryPull)
```
yes, well, therein lies the rub: if I make it `inlets`, then tests start to fail :-/

laszlovandenhoek commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -960,8 +960,8 @@ object ZipWithN {
-  def out = shape.out
-  val inSeq = shape.inSeq
+  def out: Outlet[O] = shape.out
+  val inSeq: immutable.IndexedSeq[Inlet[A]] = shape.inlets.toIndexedSeq
```
needed for binary compatibility since `inSeq` is public. I think `toIndexedSeq` is preferable to `asInstanceOf`.

johanandren replied:
If we know it is true, `asInstanceOf` is better because it should compile down to a no-op while `toIndexedSeq` is a method call.

laszlovandenhoek committed:
process additional review comments

laszlovandenhoek committed:
make inArray fully private everywhere

laszlovandenhoek committed:
Merge branch 'master' into uniform-faninout-api

laszlovandenhoek commented:
OK, so the current statet is with the internal `Array` copy inside `FanInShape`, fully private. This is a version I can get behind, but I'm currently running a benchmark simplified from `akka.stream.GraphBuilderBenchmark.graph_with_junctions()` that actually calls `in(id)` with `id > 1`, as opposed to the original. I'll do another run with the `Array` taken out and just using the `inlets` `Vector` directly, and see if the added Array initialization pays for itself in terms of faster performance of `in(id)`. Will keep you posted.

laszlovandenhoek commented:
with `Array` duplicating `Vector`:

```
[info] Benchmark                                             (complexity)   Mode  Cnt       Score      Error  Units
[info] GraphBuilderBenchmark.graph_with_junctions_immediate             1  thrpt  200  119108,988 ▒ 1840,384  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate            10  thrpt  200   34529,891 ▒  251,638  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate           100  thrpt  200    2847,236 ▒   29,108  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate          1000  thrpt  200      43,150 ▒    0,555  ops/s
```

without `Array`, using `Vector` directly:

```
[info] Benchmark                                             (complexity)   Mode  Cnt       Score     Error  Units
[info] GraphBuilderBenchmark.graph_with_junctions_immediate             1  thrpt  200  126597,255 ▒ 575,778  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate            10  thrpt  200   35090,834 ▒ 219,136  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate           100  thrpt  200    2582,587 ▒  30,460  ops/s
[info] GraphBuilderBenchmark.graph_with_junctions_immediate          1000  thrpt  200      43,732 ▒   1,062  ops/s
```

Up to `n=10` the performance with the `Array` is worse, probably due to the extra time taken creating the `Array`. My gut feeling is that this covers the majority of use cases, so I would propose to take out the `Array`. Should we make a `private lazy inArr` that is only used in `in(id)` if `n > c`, and then benchmark to find the optimum value of `c`? We already know `10 < c < 100` (if you call `in(id)` once for every input). I can't explain why `n=1000` is performing only marginally better. If this is not an artefact I think we should keep things simple and stay with the `Array`-less version.

laszlovandenhoek committed:
add benchmark heavy on the use of FanInShape.in()

laszlovandenhoek committed:
benchmark says: do not use Array for most n

johanandren commented on akka-stream/src/main/boilerplate/akka/stream/FanOutShape.scala.template:
```diff
@@ -27,8 +27,8 @@ abstract class FanOutShape[-I] private (_in: Inlet[I @uncheckedVariance], _regis
   final def in: Inlet[I @uncheckedVariance] = _in
-  final override def outlets: immutable.Seq[Outlet[_]] = _outlets
-  final override def inlets: immutable.Seq[Inlet[_]] = in :: Nil
+  override def outlets: immutable.Seq[Outlet[_]] = _outlets
```
Please also add a scaladoc here saying "Not meant for overriding outside of Akka"

johanandren reviewed (commented):
One comment. I think it's lookin' good.

johanandren commented:
Sounds weird that it should take extra time creating the array, the vector also creates arrays internally. Could you profile the benchmark and see what it is that takes time?

laszlovandenhoek commented:
Not weird. The `Vector` gets created regardless, the difference is `UniversalFanInShape` creating an _extra_ `Array` just for the `in(id)` lookups.

laszlovandenhoek committed:
add JavaDoc to un-final'ed defs to not override

laszlovandenhoek committed:
change deprecated val to def; do not use it

johanandren commented:
Ah, sorry it was about the _extra_ array, not about _only_ using an array.

laszlovandenhoek commented:
I think I covered all the review comments then. Does anyone feel that we should pursue adding a lazy Array to `UniversalFanInShape` that is used starting from a certain number of inlets, as I described earlier in https://github.com/akka/akka/pull/23321#issuecomment-315351487? If not, then I guess we're done here.

johanandren reviewed (approved):
LGTM, thanks for benching as well!

laszlovandenhoek commented:
@ktoso LGTY, too?

patriknw commented on akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala:
```diff
@@ -1292,15 +1292,15 @@ object GraphDSL extends GraphApply {
     @tailrec
     private[stream] def findOut[I, O](b: Builder[_], junction: UniformFanOutShape[I, O], n: Int): Outlet[O] = {
-      if (n == junction.outArray.length)
+      if (n == junction.outlets.length)
```
what type is the `outlets` here? I think it does `_out :: Nil` (if that is the same outlet? `length` on List would not be good, unless it's only single element lists

laszlovandenhoek replied:
`outlets` is always a `Vector`, again due to an implementation detail of `FanOutShape`.

The `_out :: Nil` that you're referring to is on `FanInShape`, so that's not relevant here.

patriknw commented on akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template:
```diff
@@ -54,28 +56,42 @@ object UniformFanInShape {
+
+  @deprecated("Use `inlets` or `in(id)` instead.", "2.5.4")
+  lazy val inSeq: immutable.IndexedSeq[Inlet[T @uncheckedVariance]] = inlets.toIndexedSeq
+  def in(n: Int): Inlet[T @uncheckedVariance] = inlets(n)
```
is it know that `inlets` is array or Vector, so we don't do this on List?

laszlovandenhoek replied:
Yes, it's always a `Vector`, although this relies on the private implementation of its superclass. With binary compatibility being the millstone that it is, I think adding a comment to that effect is the best we can do to prevent someone from inadvertently changing it.

patriknw commented on akka-stream/src/main/boilerplate/akka/stream/FanInShape.scala.template:
```diff
@@ -54,28 +56,42 @@ object UniformFanInShape {
   def in(n: Int): Inlet[T1 @uncheckedVariance] = {
     require(n > 0, "n must be > 0")
-    in1Seq(n - 1)
+    inlets(n).asInstanceOf[Inlet[T1]]
```
`inlets` is not List?

laszlovandenhoek replied:
It's a `Vector`.

patriknw reviewed (commented):
Great to cleanup this mess. Somewhat scary since the concrete types of some of the indexed access are not known. What could we do to make sure that we don't introduce a performance regression?

Perhaps you could try to add some temporary assertions in the code such as `assert(!outlets.isInstanceOf[List])` and run all tests.

laszlovandenhoek committed:
process patriknw's review comments

laszlovandenhoek commented:
Thanks for reviewing, @patriknw!. With regards to your performance concerns: the only way `inlets` or `outlets` could turn into anything else than a `Vector` is if someone changed it at the source, i.e. in `FanInShape` or `FanOutShape`. I have added comments telling people not to do so.

Nevertheless, I have run all tests with an `assert(![in|out]lets.isInstanceOf[List[_]])` at the top of `UniformFanXShape`, and all tests succeeded (phew 😉).

Finally, I have updated all `2.5.4` deprecation warnings to `2.5.5` (since that would be the tentative version for this change), and changed one `Array.size` call to `.length` instead, to save an implicit `SeqLike` conversion (thanks IntelliJ!).

patriknw reviewed (approved):
LGTM, thanks for doing the additional tests

jrudolph reviewed (commented):
I haven't read all the history of this one so don't take my word on this one.

Imo the API change is definitely a good one. I would have felt better if the implementation wouldn't have to change as well. Overall I think something is still needlessly complex with those classes (not your fault).

Overall a :+1: from me as the API changes (which are most important) seem reasonable.

patriknw commented:
Thanks a lot @laszlovandenhoek
